projects/uitdatabank/docs/search-api/advanced-queries.md

```diff
@@ -0,0 +1,1020 @@
+
+> Note that if an optional field is left empty in all languages, a translation can still be considered to be complete if all other fields that have a value are translated.
+
+**Applicable on endpoints**
+
+`/events` `/places` `/offers` `/organizers`
+
+**Possible values**
+
+Allowed values for both are any of the two-letter language ISO 639-1 codes, but the most used languages are: `nl` `fr` `de` `en`
+
+**Example**
+
+Retrieve all organizers that are fully translated in English:
+```
+GET /oragnizers/?q=completedLanguages:en
+```
+
+### created
+
+UiTdatabank keeps track of the creation dates of each document. This metadata for the creation date can be queried with the `created` parameter.
+
+**Applicable on endpoints**
+
+`/events` `/places` `/offers` `/organizers`
+
+**Possible values**
+
+Any ISO-8601 datetime, e.g. `2023-04-01T12:08:01+01:00`
+
+<!-- theme: info -->
+> Note that the `+` sign should be encoded for URLs (as `%2B`). Otherwise it will be interpreted as whitespace and the given date time will be considered invalid.
+
+**Example**
+
+Get all events and places that are created since the start of 2022:
+```
+GET /offers/?q=created:[2022-01-01T00\:00\:00%2B01\:00 TO *]
+```
+
+### creator
+
+Use the `creator` parameter to search for documents created by a specific creator.
+
+<!-- theme: warning --> 
+> At the time of writing, creator can contain a nickname or emailaddress. In the near future creator will only contain the uuid.
+
+**Applicable on endpoints**
+
+`/events` `/places` `/offers` `/organizers`
+
+**Possible values**
+
+Any uuid of a creator.
+
+**Example**
+
+Retrieve all events and places craeted by user with uuid `86a02c65-696d-5b12-a9b1-9e3bc8e6303c`:
+
+```
+GET /offers/?q=creator:86a02c65-696d-5b12-a9b1-9e3bc8e6303c
+```
+
+### dateRange
+
+Using the `dateRange` parameter, you can get all events and places that took place in a given range in the past, or will take place in a given range in the future.
+
+> By default, the search API will only return results that are currently available. In order to also retrieve results that are not available (yet), you'll need to disable the default filters for `availability`. You can reset this default as described in the Default Filters (TO DO: link to guide).
+
+**Applicable on endpoints** 
+
+`/events` `/places` `/offers` 
+
+**Possible values**
+
+Any ISO-8601 datetime, e.g. `2023-04-01T12:08:01+01:00`
+
+<!-- theme: info -->
+> Note that the `+` sign should be encoded for URLs (as `%2B`). Otherwise it will be interpreted as whitespace and the given date time will be considered invalid.
+
+**Example**
+
+Retrieve all events that took place in the year 2020:
+```
+GET /events/?q=dateRange:[2020-01-01T00\:00\:00%2B01\:00 TO 2021-12-31T23\:59\:59%2B01\:00]&availableTo=*&availableFrom=*
+```
+
+### \_exists_
+
+With the `_exists_` parameter you can search for the documents that have a specific property in their JSON body.
```

Comment:
Technically it returns documents that have the given field name in their index. Which is a bit different, because we do not index all JSON fields, or we index them with a slightly different name.

For example all events have a `sameAs` property with a link to UiTinVlaanderen, but we don't index that so `_exists_:sameAs` will return 0 results.

And the reverse is also true: `_exists_:address` will return places as well as events, while events never have an `address` property in their JSON. Because we index an event's location's address as `address`, so you can search on `address` in both events and places at the same time

Reply:
If it helps to visualize it, you can find the properties that we index for events here: https://github.com/cultuurnet/udb3-search-service/blob/main/src/ElasticSearch/Operations/json/mapping_event.json

So what we do is we take the UDB JSON, then transform it to a new JSON document that has all or some of the properties listed in this mapping file. The mapping file is then used to tell Elasticsearch how to analyze specific fields. (Not how to transform the UDB JSON to the JSON documents in Elasticsearch!)

Basically what `_exists_` then does is check if a JSON document in Elasticsearch has one of these properties from the mapping file.

And it's also these field names that you can filter on in the advanced queries, if their analyzer supports it. (For example you cannot filter on `geo` or `geo_point` in the advanced queries because the `geo_shape` and `geo_point` field types are not supported in "query string queries" (= Elasticsearch name for what we call "advanced queries")

Here are the ones for places and organizers:

- https://github.com/cultuurnet/udb3-search-service/blob/main/src/ElasticSearch/Operations/json/mapping_place.json
- https://github.com/cultuurnet/udb3-search-service/blob/main/src/ElasticSearch/Operations/json/mapping_organizer.json